src/resdk/resources/metadata.py

```diff
@@ -0,0 +1,247 @@
+            )
+
+    def get_df(self, parser=None, **kwargs):
+        """Get table as pd.DataFrame."""
+        # Do not use cached value if parser is specified
+        if self._df is None or parser is not None:
+            if self.id is None:
+                return None
+            if not self.output or "table" not in self.output:
+                raise ValueError('Cannot parse, no output with name "table".')
+
+            # Enable parsing the byte stream with arbitrary parser, not just pandas
+            # Otherwise try to guess the parser based on file extension
+            basename = self.output["table"]["file"]
+            if parser is None:
+                if basename.endswith("xls"):
+                    parser = pd.read_excel
+                    kwargs = dict(engine="xlrd")
+                elif basename.endswith("xlsx"):
+                    parser = pd.read_excel
+                    kwargs = dict(engine="openpyxl")
+                elif any(basename.endswith(ext) for ext in ["tab", "tsv"]):
+                    parser = pd.read_csv
+                    kwargs = dict(
+                        sep="\t", low_memory=False, float_precision="round_trip"
+                    )
+                else:
+                    parser = pd.read_csv
+                    kwargs = dict(low_memory=False, float_precision="round_trip")
+
+            df = parser(self.df_bytes, **kwargs)
+
+            self.validate_df(df)
+            self._df = df
+
+        return self._df
+
+    def set_df(self, value):
+        """Set df."""
+        if self.id:
+            raise ValueError(
+                "Setting df attribute on already uploaded Metadata is not allowed."
+            )
+        if not self.collection:
+            # Validation is not possible without collection
+            raise ValueError(
+                "Setting df attribute before setting collection is not allowed."
+            )
+
+        self.validate_df(value)
+        self._df = value
+
+    df = property(get_df, set_df)
+
+    def save(self):
+        """Save Metadata to the server.
+
+        If Metadata is already uploaded: update. Otherwise, create new one.
+        """
+        if self.id:
+            super().save()
+        else:
+            if not self.collection:
+                raise ValueError("Collection must be set before saving.")
+            if self.df is None or self.df.empty:
+                raise ValueError("Attribute df must be set before saving.")
+
+            # All resdk machinery for uploading files works with real
+            # files on the system. Ideally we would support "file upload"
+            # from a stream, but for now, let's use a tempfile solution.
+            tmp_file = tempfile.NamedTemporaryFile(suffix=".csv")
+            self.df.to_csv(tmp_file.name, index=False)
+            inputs = self.resolwe._process_inputs({"src": tmp_file.name}, self.process)
+            tmp_file.close()

```

Comment:
@robertcv Can you weigh in here? In the first line, a temporary file is created and opened for writing. In the next line, the same file is opened again for writing by `pandas` using a different handler. Does this work on all operating systems, i.e., no locking of files opened for writing? After writing, `pandas`  closes the file, which is where the flush occurs as well, so we have a guarantee to have the full file content available for upload in the next line. If we now wrote something to the same file with `tmp_file.write(b"blabla")` and flushed, the written `DataFrame` would be partially overwritten by `"blabla"`. After `self.df.to_csv(...)` the temporary file is not deleted yet even though the file has been closed by `pandas` because the deletion of temporary files is handled by `tmp_file.close()`.

Reply:
@JureZmrzlikar Perhaps `self.df.to_csv(tpm_file, index=False)` would be a better choice to avoid creating the second file handler and reopening the same file for writing, but I'm not sure if `pandas` flushes afterwards.

Reply:
Flush is now called manually to ensure all content is present in the temporary file.